File: backend/app/api/backtest.py
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel
from app.services.backtest_service import run_backtest

router = APIRouter()


class StrategyParams(BaseModel):
    score_rising_days: int = 3
    oversold_std_mult: float = 2.0
    take_profit_pct: float = 0.15
    take_profit_std_mult: float = 1.5
    stop_loss_pct: float = 0.05
    add_position_pct: float = 0.05
    half_position_ratio: float = 0.5
    # 三因子计算参数
    bias_n: int = 20
    momentum_day: int = 20
    slope_n: int = 20
    efficiency_n: int = 20
    zscore_window: int = 60
    # SuperTrend+MA 参数
    recent_high_window: int = 25


class BacktestRequest(BaseModel):
    stock_code: str
    start_date: str
    end_date: str
    initial_cash: float = 100000.0
    strategy_name: str = "three_factors"
    strategy_params: StrategyParams = StrategyParams()


@router.post("/run")
async def run_backtest_endpoint(request: BacktestRequest):
    try:
        return run_backtest(
            stock_code=request.stock_code,
            start_date=request.start_date,
            end_date=request.end_date,
            initial_cash=request.initial_cash,
            strategy_name=request.strategy_name,
            strategy_params=request.strategy_params.model_dump(),
        )
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        import traceback
        traceback.print_exc()
        raise HTTPException(status_code=500, detail=f"回测失败: {e}")

File: frontend/src/types/backtest.ts
export interface StrategyParams {
  score_rising_days: number
  oversold_std_mult: number
  take_profit_pct: number
  take_profit_std_mult: number
  stop_loss_pct: number
  add_position_pct: number
  half_position_ratio: number
  // 三因子计算参数
  bias_n: number
  momentum_day: number
  slope_n: number
  efficiency_n: number
  zscore_window: number
}

export const DEFAULT_STRATEGY_PARAMS: StrategyParams = {
  score_rising_days: 3,
  oversold_std_mult: 2.0,
  take_profit_pct: 0.15,
  take_profit_std_mult: 1.5,
  stop_loss_pct: 0.05,
  add_position_pct: 0.05,
  half_position_ratio: 0.5,
  bias_n: 20,
  momentum_day: 20,
  slope_n: 20,
  efficiency_n: 20,
  zscore_window: 60,
}

export interface Trade {
  date: string
  direction: '买入' | '卖出'
  reason: string
  price: number
  shares: number
  amount: number
  transfer_fee: number
  commission: number
  stamp_tax: number | null
  total_fee: number
  pnl: number | null
  remaining_cash: number
}

export interface BacktestSummary {
  initial_cash: number
  final_assets: number
  total_return: number
  total_fees: number
  benchmark_return: number
  excess_return: number
  final_shares: number
  final_cash: number
}

export interface BacktestResponse {
  success: boolean
  dates: string[]
  open: (number | null)[]
  high: (number | null)[]
  low: (number | null)[]
  close: (number | null)[]
  ma5: (number | null)[]
  ma20: (number | null)[]
  ma60: (number | null)[]
  supertrend: (number | null)[]
  supertrend_direction: (number | null)[]
  score: (number | null)[]
  score_mean: (number | null)[]
  score_std: (number | null)[]
  trades: Trade[]
  summary: BacktestSummary
}

export interface BacktestRequest {
  stock_code: string
  start_date: string
  end_date: string
  initial_cash: number
  strategy_name: string
  strategy_params: StrategyParams
}
